arch/xtensa/core/userspace.S

```diff
@@ -20,6 +22,37 @@
 .global	xtensa_do_syscall
 .align	4
 xtensa_do_syscall:
+#if XCHAL_HAVE_THREADPTR == 0
+	wsr a2, ZSR_A2SAVE
+	rsync
+
+	movi a0, xtensa_is_user_context_epc
+	rsr.epc1 a2
+	bne a0, a2, _not_checking_user_context
+
+	addi a2, a2, 3
+	wsr.epc1 a2
+
+	movi a0, PS_RING_MASK
+	rsr.ps a2
+	and a2, a2, a0
+
+	/* Need to set return to 1 if RING != 0,
+	 * so we won't be leaking which ring we are in
+	 * right now.
+	 */
+	beqz a2, _is_user_context_return
+
+	movi a2, 1
```

Comment:
We could simplify this logic having this logic having this exception handled in User/Kernel vectors and setting UM bit properly. I have a prototype doing it but without the  `EPC1` trick. This is very clever.

Reply:
I thought about it... but then all the existing code will need to be audited to make sure they work on both non-UM/UM modes. So for now, doing this here to keep the PR a little bit simpler.